From e37231e2c8107ce126da0bfe1b9bded6710f425a Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Sat, 2 Dec 2023 16:59:38 +0100
Subject: [PATCH 0802/1141] media: dcmipp: isp: correct src pad format setting

Adjust the source pad format whenever setting the sink pad of the
dcmipp-isp subdev.

Change-Id: I28015a1d2ff533d920544227dc8a381983126420
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/344610
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-isp.c        | 25 ++++++++++++++-----
 1 file changed, 19 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
index 85972feafef9..fd133eb3ee3c 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
@@ -603,16 +603,29 @@ static int dcmipp_isp_set_fmt(struct v4l2_subdev *sd,
 
 	dcmipp_isp_adjust_fmt(&fmt->format, fmt->pad);
 
-	if (IS_SINK(fmt->pad) && fmt->which == V4L2_SUBDEV_FORMAT_ACTIVE) {
-		/* When setting sink format, we have to update the src format */
-		isp->src_fmt = fmt->format;
+	/* When setting sink format, we have to update the src format */
+	if (IS_SINK(fmt->pad)) {
+		struct v4l2_mbus_framefmt *src_pad_fmt;
+
+		if (fmt->which == V4L2_SUBDEV_FORMAT_ACTIVE)
+			src_pad_fmt = &isp->src_fmt;
+		else
+			src_pad_fmt = v4l2_subdev_get_try_format(sd, state, 1);
+
+		*src_pad_fmt = fmt->format;
 		if (fmt->format.code >= MEDIA_BUS_FMT_Y8_1X8 &&
 		    fmt->format.code < MEDIA_BUS_FMT_SBGGR8_1X8)
-			isp->src_fmt.code = MEDIA_BUS_FMT_YUV8_1X24;
+			src_pad_fmt->code = MEDIA_BUS_FMT_YUV8_1X24;
 		else
-			isp->src_fmt.code = MEDIA_BUS_FMT_RGB888_1X24;
+			src_pad_fmt->code = MEDIA_BUS_FMT_RGB888_1X24;
+
+		dev_dbg(isp->dev, "%s: source format update: new:%dx%d (0x%x, %d, %d, %d, %d)\n",
+			isp->sd.name,
+			src_pad_fmt->width, src_pad_fmt->height,
+			src_pad_fmt->code, src_pad_fmt->colorspace,
+			src_pad_fmt->quantization,
+			src_pad_fmt->xfer_func, src_pad_fmt->ycbcr_enc);
 	}
-	/* TODO - we need to update the try src format as well */
 
 	dev_dbg(isp->dev, "%s: %s format update: old:%dx%d (0x%x, %d, %d, %d, %d) new:%dx%d (0x%x, %d, %d, %d, %d)\n",
 		isp->sd.name,
-- 
2.39.2

